### PR TITLE
FIX #15739 dynamic price

### LINKED ISSUE #15739: Bug : product selling price not refreshed on product list when the pricing mode is set to be computed
# Bug
Product selling price not refreshed, the first time, on product list when the pricing mode is set to be computed.

## Environment
-Version: 12.0.3
-OS: Linux Debian 3.16.64-2 (2019-04-01) x86_64
-Web server: Apache/2
-PHP: PHP 5.6.40
-Database: MySQL or MariaDB 5.6.48-log
-URL(s) /product/price.php?id=13371&action=edit_price
- **Dynamically calculated prices module activated**

## Expected and actual behavior
Modifying a product selling price by setting it to be computed, for example  price=#global_coeff3#*#supplier_min_price#, after the first registration of the product, returning to the list the selling price is not the same. If you return to the product and reopen the selling price and then saving without changing nothing, the selling price will be good this time.

**After a survey of users, the problem does not seem to be systematic but reproducible in about 10 to 20% of cases.**

## Steps to reproduce the behavior
Locating a product whose selling price differs from that of its peers:
image
Let's see what happened to his selling price:
image
Bizarre, everything seems correct, so let's change this selling price:
image
Everything seems correct, the selling price is calculated using a simple formula (X_coeff3 corresponds to price=#global_coeff3#*#supplier_min_price#) involving the purchase price and also includes a minimum price.
I save, in this state, without modifying anything, the history is enriched by a line identical to the previous one:
image
Then I go back to the list :
image
And then, oh surprise, everything is back in order, the selling price is in line with what was calculated the first time.

Isn't there a variable badly initialized somewhere that would do its job badly?

Discussion on french forum[ here](https://www.dolibarr.fr/forum/t/calcul-de-prix-dynamique/24377).

### PR DESCRIPTION
**Pull request by atm-florianm**
# Fix #15739 
A dynamic (selling) price is computed using an expression that can rely on several values, including minimum selling price, best buying price, weight, etc.

When the `Product` object is fetched, the price is automatically recomputed, which explains why the price displayed on a product card is always up to date. However, when it is recomputed, it is not always stored, so pages that rely on SQL queries to get a product's selling price (such as the product list) show a potentially outdated price value.

The price of a product using a dynamic price should be updated (in the database) whenever one of the values used its price expression (or the expression itself) changes.

This PR is not complete in that respect, but it lays the foundation for further fixes.

### CODE REVIEW

File: htdocs/fourn/class/fournisseur.product.class.php [hunk 1]
```diff
@@ -383,6 +383,12 @@ public function update_buyprice($qty, $buyprice, $user, $price_base_type, $fourn
 				if (empty($error))
 				{
 					$this->db->commit();
+					if (!empty($conf->dynamicprices->enabled) && !empty($this->fk_price_expression)) {
+						// Computed selling prices may depend on the purchase price. By re-fetching the product, we
+						// force it to re-compute the selling price and to save it if it has changed.
+						$product = new Product($this->db);
+						if ($product->fetch($this->id) > 0) $product->updatePriceFromExpression();
```
Comment (eldy) [inline htdocs/fourn/class/fournisseur.product.class.php:390]: Can you confirm you need to create a new object Product.
Why not just $this->updatePriceFromExpression()

File: htdocs/product/price.php [hunk 2]
```diff
@@ -378,6 +378,7 @@
 			$action = '';
 			setEventMessages($langs->trans("RecordSaved"), null, 'mesgs');
 			$db->commit();
+			$object->updatePriceFromExpression();
```
Comment (eldy) [inline htdocs/product/price.php:381]: This is outside the transaction ?  After the commit ? Why ?

File: htdocs/fourn/class/fournisseur.product.class.php (hunk 1 above)
Comment (atm-florianm) [reply htdocs/fourn/class/fournisseur.product.class.php]: Sorry, this was the remnant of an older code layout which required it, but it is no longer necessary. I modified the code accordingly.

File: htdocs/product/price.php (hunk 2 above)
Comment (atm-florianm) [reply htdocs/product/price.php]: OK

**Comment by eldy** [comment]
Because this is an experimental feature, we have no reason to fix it in v12. Becasue, it seems there is some changes related to same part of code into v13, this will break v13 too. Can you try to push your PR to v13 or develop branch ? (by merging the conflict due to other changes on same files) ?

**Comment by eldy** [comment]
This PR is now too old, so i close it. A lot of code may have been validated since the PR and the PR may be obsolete, or too difficult to merge due to other changes. Best choice is too cancel it to avoid to keep PR with too low chances to be validated.

**Comment by MarcVJ** [comment]
Hello,

Would it be possible to reopen this fix and try to add it to a future Dolibarr version ?

The bug is still present on Dolibarr v20.0.2. 
The formula we are using is :
price = #supplier_min_price_with_discount# * #global_coeff1# 
We tried other formulas and it’s the same.
We always have to go on the selling price page of the product, edit it and save it without touching anything, for the price to actually update in the product list.
And we have to do this on every product where the supplier price is changed.

We would greatly appreciate an update solving this problem !

Regards,

**Pull request closed without merging**